packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -559,135 +559,292 @@ impl<'m> RunnerImpl<'m> {
+        target_len: usize,
+    ) -> (Buffer, usize) {
+        // TODO: OPTIM: We currently copy the buffers because the JavaScript representation of
+        //   arrays does not match the Rust implementation. Try to reduce copies where possible by
+        //   reusing it, i.e. check, if `target_len` <= `capacity` and constructing it from raw
+        //   parts.
+
+        // For each value in the buffer, we have a start offset and an end offset. The start offset
+        // is equal to the end offset of the previous value, thus we need `num_values + 1`
+        // offset values.
+        let mut builder = BufferBuilder::new(target_len + 1);
+
+        let offsets = slice::from_raw_parts(ptr.as_ptr(), len + 1);
+        debug_assert_eq!(offsets[0], 0, "Offset buffer does not start with `0`");
+        debug_assert!(
+            offsets.iter().all(|o| *o >= 0),
+            "Offset buffer contains negative values"
+        );
+        debug_assert!(offsets.is_sorted(), "Offsets are not ordered");
+
+        // Read data from JS
+        builder.append_slice(offsets);
+
+        let last = offsets[len];
+
+        // Ensure we don't subtract a larger number from a smaller
+        // TODO: Use `buffer.resize()` instead of `builder.append_n()`
+        debug_assert!(
+            target_len >= len,
+            "Expected offset count is smaller than the actual buffer: {:?}",
+            slice::from_raw_parts(ptr.as_ptr(), len + 1)
+        );
+        builder.append_n(target_len - len, last);
+        (builder.finish(), last as usize)
+    }
+
+    /// Adds a new packed buffer from the provided `ptr` and `capacity` with at least a capacity
+    /// of `target_len` elements and `len` elements are copied from `ptr`.
+    ///
+    /// # SAFETY
+    ///
+    /// - `ptr` must be valid for `len` reads of `T`
+    unsafe fn read_boolean_buffer(
+        &self,
+        ptr: NonNull<u8>,
+        len: usize,
+        _capacity: usize, // for future use to create a `Buffer::from_raw_parts`
+        target_len: usize,
+    ) -> Buffer {
+        // TODO: OPTIM: We currently copy the buffers because the JavaScript representation of
+        //   arrays does not match the Rust implementation. Try to reduce copies where possible by
+        //   reusing it, i.e. check, if `target_len` <= `capacity` and constructing it from raw
+        //   parts.
+        // Create a buffer for `target_len` elements
+        let mut builder = BooleanBufferBuilder::new(target_len);
+
+        // Read data from JS
+        builder.append_packed_range(
+            0..len,
+            slice::from_raw_parts(ptr.as_ptr(), bit_util::ceil(len, 8)),
+        );
+        builder.resize(target_len);
+        builder.finish()
+    }
+
     /// TODO: DOC, flushing from a single column
     fn array_data_from_js(
         &mut self,
         mv8: &'m MiniV8,
         data: &mv8::Value<'m>,
-        field: &Field,
+        data_type: &DataType,
```

Comment:
Was it not always possible to pass the whole field?

Reply:
It's not required to pass the field

Reply:
It's just that for logging it's nice to have the field name

Reply:
We currently don't use logging inside of this function. Also, it was `data_type` before upgrading Arrow.

Reply:
Fair enough, I meant in the error message